src/lib.rs

```diff
@@ -237,10 +265,25 @@ impl Message {
         }
     }
 
-    pub fn verify(&self, sig: [u8; 65]) -> Result<Vec<u8>, VerificationError> {
-        if !self.valid_now() {
+    pub fn verify(
+        &self,
+        sig: [u8; 65],
+        domain: Option<&Authority>,
+        nonce: Option<&str>,
+        timestamp: Option<&DateTime<Utc>>,
+    ) -> Result<Vec<u8>, VerificationError> {
+        if match timestamp {
+            None => !self.valid_now(),
+            Some(timestamp) => !self.valid_at(timestamp),
+        } {
             Err(VerificationError::Time)
         } else {
+            if domain.is_some() && *domain.unwrap() != self.domain {
+                return Err(VerificationError::DomainMismatch);
+            }
+            if nonce.is_some() && *nonce.unwrap() != self.nonce {
+                return Err(VerificationError::NonceMismatch);
+            }
```

Comment:
I believe this doesn't work if the domain matches but the nonce does not (I believe it would report that it was valid).

I think this sort of approach would be better:
```
        match domain {
            Some(d) if *d != self.domain => return Err(VerificationError::DomainMismatch),
            _ => (),
        };

        match nonce {
            Some(n) if n != self.nonce => return Err(VerificationError::NonceMismatch),
            _ => (),
        };

        match timestamp {
            Some(t) if !self.valid_at(t) => return Err(VerificationError::Time),
            _ => (),
        };
```

I'm was taking a look at problems with the `siwe-ex` verification and I think this might be a contributing factor.